File: 450/Ratndeep/Binary Tree/Reverse level order.py
class Node:
    def __init__(self,data):
        self.data=data 
        self.left=None 
        self.right=None 

class BTree:
    def __init__(self):
        self.root=None 


    def addNode(self,data): 
        node = Node(data) 

        while(root):
            if root.data>node.data:
                if root.right==None:
                    root.right = node 
                    break 
                else: 
                    root=root.right 
            else: 
                if root.left==None: 
                    root.left=node 
                    break 
                else: 
                    root=root.left 
    
    def reverseLevelOrder(self,root):
   
        stack=[]
        queue=[root]
        while queue!=[]:
            stack.insert(0,queue[0].data)
            cur_root=queue.pop(0)
            if cur_root.right:
                queue.append(cur_root.right)
            if cur_root.left:
                queue.append(cur_root.left)
        return stack

bt = BTree() 
data = list(map(int,input().split()))
for i in range(len(data)):
    bt.addNode(data[i]) 
print(*bt.reverseLevelOrder(bt.root))

File: 450/Utkarsh/binaryTree/reverseLevelorder.cpp
//https://practice.geeksforgeeks.org/problems/reverse-level-order-traversal/1#

#include<bits/stdc++.h>
using namespace std;

/*   
struct Node
{
    int data;
    Node* left;
    Node* right;
}; */


vector<int> reverseLevelOrder(Node *root)
{
  vector<int > v;
  queue < Node* > q;
  stack < Node* > s;
  q.push(root);
  while(!q.empty()){
      Node * temp = q.front(); q.pop();
    //   cout<< temp->data<<" ";
      s.push(temp);
      if(temp->right)       q.push(temp->right);
      if (temp->left)  q.push(temp->left);
  }
  while(!s.empty()){
      Node * temp = s.top();
      s.pop();
      if(temp)
      v.push_back(temp->data);
  }
  return v;
}